extension/api.js

```diff
@@ -0,0 +1,195 @@
+        },
+
+        // `onSearchEngineModified` is an event that occurs when the list of
+        // search engines has changed, e.g., a new engine has been added or an
+        // engine has been removed.
+        //
+        // See: https://searchfox.org/mozilla-central/source/toolkit/components/search/SearchUtils.jsm#145-152
+        onSearchEngineModified: new ExtensionCommon.EventManager({
+          context,
+          name: "addonsSearchExperiment.onSearchEngineModified",
+          register: (fire) => {
+            const onSearchEngineModifiedObserver = (
+              aSubject,
+              aTopic,
+              aData
+            ) => {
+              if (
+                aTopic !== SEARCH_TOPIC_ENGINE_MODIFIED ||
+                // We are only interested in these modified types.
+                !["engine-added", "engine-removed", "engine-changed"].includes(
+                  aData
+                )
+              ) {
+                return;
+              }
+
+              fire.async();
+            };
+
+            Services.obs.addObserver(
+              onSearchEngineModifiedObserver,
+              SEARCH_TOPIC_ENGINE_MODIFIED
+            );
+
+            return () => {
+              Services.obs.removeObserver(
+                onSearchEngineModifiedObserver,
+                SEARCH_TOPIC_ENGINE_MODIFIED
+              );
+            };
+          },
+        }).api(),
+
+        // `onBeforeRedirect` is an event very similar to the public web
+        // request `onBeforeRedirect` but it is registered in the privileged
+        // code to make sure that we can read properties on a valid channel
+        // (wrapper). The registered listeners will received the `addonId` in
+        // addition to these props: `requestId`, `url` and `redirectUrl`.
+        onBeforeRedirect: new ExtensionCommon.EventManager({
+          context,
+          name: "addonsSearchExperiment.onBeforeRedirect",
+          register: (fire, filter) => {
+            const listener = ({ requestId, url, redirectUrl }) => {
+              const wrapper = ChannelWrapper.getRegisteredChannel(
+                requestId,
+                context.extension.policy,
+                context.xulBrowser.frameLoader.remoteTab
+              );
+
+              // When we detect a redirect, we read the request property,
+              // hoping to find an add-on ID corresponding to the add-on that
+              // initiated the redirect. It might not return anything when the
+              // redirect is a search server-side redirect but it can also be
+              // caused by an error.
+              let addonId;
+
+              try {
+                addonId = wrapper?.channel
+                  ?.QueryInterface(Ci.nsIPropertyBag)
+                  ?.getProperty("redirectedByExtension");
+              } catch (err) {
+                Cu.reportError(err);
+              }
+
+              fire.async({ addonId, redirectUrl, requestId, url });
```

Comment:
Make this `fire.sync`, so that the event is not unnecessarily delayed, which could potentially result in out-of-order events relative to the `fire.sync` call from the `webRequest.onBeforeRequest` event.